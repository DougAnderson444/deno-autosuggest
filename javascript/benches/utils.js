
function first2digits(num) {
    let scale = 1
    while (num >= 100) {
        scale *= 10
        num   /= 10
    }
    return Math.round(num) * scale
}

exports.first2digits = first2digits
