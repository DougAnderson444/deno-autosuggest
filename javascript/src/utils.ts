import type {Word} from './tokenization/word'
import type {HighlightedTextChunk} from './index'


export function once<T>(fn: () => T): () => T {
    let result: T | undefined = undefined
    return () => {
        if (result !== undefined) {
            return result
        } else {
            result = fn()
            return result
        }
    }
}


export function toChunks(title: string): HighlightedTextChunk[] {
    const split  = title.split(/{{|}}/g)
    const chunks = []
    for (let i = 0; i < split.length; i++) {
        if (split[i] != '') {
            chunks.push({
                text: split[i],
                highlight: i % 2 === 1,
            })
        }
    }
    return chunks
}


export function serializeWords(words: Word[]): Uint32Array {
    const serialized = new Uint32Array(words.length * 6)
    for (let i = 0; i < words.length; i++) {
        const word = words[i]
        serialized[i * 6 + 0] = word.offset
        serialized[i * 6 + 1] = word.slice[0]
        serialized[i * 6 + 2] = word.slice[1]
        serialized[i * 6 + 3] = word.stem
        serialized[i * 6 + 4] = word.pos
        serialized[i * 6 + 5] = +word.fin
    }
    return serialized
}