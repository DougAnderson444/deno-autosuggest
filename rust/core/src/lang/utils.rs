use std::collections::HashMap;


pub fn compile_utf_map(static_map: &[(&str, &str)]) -> HashMap<Vec<char>, Vec<char>> {
    let mut compiled_map = HashMap::with_capacity(static_map.len());
    for (pattern, replace) in static_map {
        compiled_map.insert(
            pattern.chars().collect(),
            replace.chars().collect()
        );
    }
    compiled_map
}
