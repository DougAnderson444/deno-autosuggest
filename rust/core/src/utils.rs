#![macro_use]

use std::cell::RefCell;


macro_rules! min {
    ($x: expr) => ($x);
    ($x: expr, $($z: expr),+) => (::std::cmp::min($x, min!($($z),*)));
}


macro_rules! max {
    ($x: expr) => ($x);
    ($x: expr, $($z: expr),+) => (::std::cmp::max($x, max!($($z),*)));
}
