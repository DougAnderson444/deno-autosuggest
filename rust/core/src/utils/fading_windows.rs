use std::cmp::min;


#[derive(Debug)]
pub struct FadingWindows<'a, T: 'a> {
    v: &'a [T],
    size: usize,
}

impl<'a, T: 'a> FadingWindows<'a, T> {
    pub fn new(v: &'a [T], size: usize) -> Self {
        if size == 0 && v.len() > 0 {
            panic!("Can't have zero window size with nonempty slice");
        }
        Self { v, size }
    }
}

impl<'a, T> Iterator for FadingWindows<'a, T> {
    type Item = &'a [T];

    #[inline]
    fn next(&mut self) -> Option<&'a [T]> {
        if self.v.len() == 0 {
            None
        } else {
            let window = Some(&self.v[.. min(self.size, self.v.len())]);
            self.v = &self.v[1..];
            window
        }
    }

    #[inline]
    fn size_hint(&self) -> (usize, Option<usize>) {
        let len = self.v.len();
        (len, Some(len))
    }
}

impl<'a, T> ExactSizeIterator for FadingWindows<'a, T> { }


#[cfg(test)]
mod tests {
    use super::FadingWindows;

    #[test]
    fn fading_windows_empty() {
        let input: [usize; 0] = [];
        for size in 0 .. 4 {
            let output = FadingWindows::new(&input, size).collect::<Vec<_>>();
            let expect: Vec<&[usize]> = Vec::new();
            assert_eq!(output, expect);
        }
    }


    #[test]
    fn fading_windows_singleton() {
        let input: [usize; 1] = [10];
        for size in 1 .. 4 {
            let output = FadingWindows::new(&input, size).collect::<Vec<_>>();
            let expect: Vec<&[usize]> = vec![&[10]];
            assert_eq!(output, expect);
        }
    }


    #[test]
    fn fading_windows_len_gt_size() {
        let size = 3;
        let input: [usize; 5] = [10, 20, 30, 40, 50];
        let expect: [&[usize]; 5] = [&[10, 20, 30], &[20, 30, 40], &[30, 40, 50], &[40, 50], &[50]];
        let output = FadingWindows::new(&input, size).collect::<Vec<_>>();
        assert_eq!(output, expect);
    }


    #[test]
    fn fading_windows_size_gt_len() {
        let size = 5;
        let input: [usize; 3] = [10, 20, 30];
        let expect: [&[usize]; 3] = [&[10, 20, 30], &[20, 30], &[30]];
        let output = FadingWindows::new(&input, size).collect::<Vec<_>>();
        assert_eq!(output, expect);
    }


    #[test]
    #[should_panic]
    fn fading_windows_zero_size() {
        let size = 0;
        let input: [usize; 3] = [10, 20, 30];
        FadingWindows::new(&input, size);
    }
}
