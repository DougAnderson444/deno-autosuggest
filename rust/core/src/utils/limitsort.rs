use std::cmp::Ordering;


pub trait LimitSort: Iterator + Sized {
    fn limit_sort<F>(self, limit: usize, sort_fn: F) -> LimitSortIter<Self::Item, Self, F> where
        F: (FnMut(&Self::Item, &Self::Item) -> Ordering)
    {
        LimitSortIter {
            sort_fn,
            source: self,
            buffer: Vec::with_capacity(limit * 2),
            limit,
            done: false,
        }
    }
}


pub struct LimitSortIter<T, I: Iterator<Item=T>, F: FnMut(&T, &T) -> Ordering> {
    sort_fn: F,
    source:  I,
    buffer:  Vec<T>,
    limit:   usize,
    done:    bool,
}


impl<T, I: Iterator<Item=T>, F: FnMut(&T, &T) -> Ordering> Iterator for LimitSortIter<T, I, F> {
    type Item = T;

    fn next(&mut self) -> Option<Self::Item> {
        let Self { buffer, source, sort_fn, done, .. } = self;
        let limit = self.limit;

        if !*done {
            while let Some(item) = source.next() {
                buffer.push(item);
                if buffer.len() >= limit * 2 {
                    buffer.sort_by(|x, y| sort_fn(x, y));
                    buffer.truncate(limit);
                }
            }
            buffer.sort_by(|x, y| sort_fn(x, y));
            buffer.truncate(limit);
            buffer.reverse();
            *done = true;
        }

        buffer.pop()
    }
}


impl<I: Iterator> LimitSort for I { }
