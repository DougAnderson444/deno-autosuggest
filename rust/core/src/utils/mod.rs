#![macro_use]

mod trigrams;
mod limitsort;
mod fading_windows;

pub use trigrams::{Trigrams, TrigramIter};
pub use limitsort::{LimitSort, LimitSortIter};
pub use fading_windows::FadingWindows;


macro_rules! min {
    ($x: expr) => ($x);
    ($x: expr, $($z: expr),+) => (::std::cmp::min($x, min!($($z),*)));
}


macro_rules! max {
    ($x: expr) => ($x);
    ($x: expr, $($z: expr),+) => (::std::cmp::max($x, max!($($z),*)));
}


pub fn to_vec<T: AsRef<str>>(chars: T) -> Vec<char> {
    chars.as_ref().chars().collect()
}


#[cfg(test)]
pub fn to_str<T: AsRef<[char]>>(chars: T) -> String {
    chars.as_ref().iter().collect()
}
