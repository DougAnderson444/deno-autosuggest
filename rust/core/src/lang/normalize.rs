use std::collections::HashMap;
use crate::utils::FadingWindows;

const NORM_MAX_PATTERN_LEN: usize = 2;


pub struct Normalize<'a> {
    windows: FadingWindows<'a, char>,
    map:     &'a HashMap<Vec<char>, Vec<char>>,
    skip:    usize,
}


impl<'a> Normalize<'a> {
    pub fn new(source: &'a [char], map: &'a HashMap<Vec<char>, Vec<char>>) -> Self {
        Self {
            windows: FadingWindows::new(source, NORM_MAX_PATTERN_LEN),
            map,
            skip: 0,
        }
    }
}


impl<'a> Iterator for Normalize<'a> {
    type Item = (&'a [char], &'a [char]);

    fn next(&mut self) -> Option<Self::Item> {
        let mut window = self.windows.next()?;

        while self.skip > 0 {
            window = self.windows.next()?;
            self.skip -= 1;
        }

        for len in (1 .. window.len() + 1).rev() {
            let pattern = &window[..len];
            if let Some(replace) = self.map.get(pattern) {
                self.skip = pattern.len() - 1;
                return Some((pattern, replace));
            }
        }

        Some((&window[..1], &window[..1]))
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        self.windows.size_hint()
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use super::super::utils::compile_utf_map;
    use super::Normalize;

    fn normalize<'a>(word: &'a [char]) -> (Vec<String>, Vec<String>) {
        let norm_map = compile_utf_map(&[
            ("ó", "o"),
            ("ó", "o"),
            ]);
        let mut source = Vec::new();
        let mut norm   = Vec::new();
        for (chunk1, chunk2) in Normalize::new(word, &norm_map) {
            source.push(chunk1.iter().collect::<String>());
            norm  .push(chunk2.iter().collect::<String>());
        }
        (source, norm)
    }

    #[test]
    fn utf_normalize_empty() {
        let input  = "".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_singleton_normal() {
        let input  = "f".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_singleton_replaced() {
        let input  = "ó".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_single_nfd_replaced() {
        let input  = "ó".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_normal() {
        let input  = "foobar".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_nfc_beginning() {
        let input  = "óoobar".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_nfd_beginning() {
        let input  = "óoobar".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_nfc_middle() {
        let input  = "foóbar".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_nfd_middle() {
        let input  = "foóbar".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_nfc_end() {
        let input  = "foobaó".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn utf_normalize_nfd_end() {
        let input  = "foobaó".chars().collect::<Vec<_>>();
        let output = normalize(&input[..]);
        assert_debug_snapshot!(output);
    }
}