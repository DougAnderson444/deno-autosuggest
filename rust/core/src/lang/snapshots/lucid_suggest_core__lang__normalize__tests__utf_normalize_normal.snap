---
source: src/lang/normalize.rs
expression: output
---
(
    [
        "f",
        "o",
        "o",
        "b",
        "a",
        "r",
    ],
    [
        "f",
        "o",
        "o",
        "b",
        "a",
        "r",
    ],
)
